_posts/2020-10-09-torch-optim/torch_network_with_optim.Rmd

```diff
@@ -0,0 +1,269 @@
+---
+title: "Optimizers in torch"
+description: > 
+ Today, we wrap up our mini-series on torch basics, adding to our toolset two abstractions we hadn't made use of before: loss functions and optimizers.
+
+author:
+  - name: Sigrid Keydana
+    affiliation: RStudio
+    affiliation_url: https://www.rstudio.com/
+slug: keydanatorchnetworkfromscratch
+date: 10-01-2020
+categories:
+  - Torch
+  - R
+bibliography: bibliography.bib
+output:
+  distill::distill_article:
+    self_contained: false
+    toc: true
+preview: images/preview.jpg
+---
+
+```{r setup, include=FALSE}
+knitr::opts_chunk$set(echo = TRUE, eval = FALSE)
+```
+
+```{r}
+
+```
+
+This is the fourth and last installment in a series introducing `torch` basics. Initially, we [focused on *tensors*](https://blogs.rstudio.com/ai/posts/2020-10-01-torch-network-from-scratch/). To illustrate their power, we coded a complete (if toy-size) neural network from scratch. We didn't make use of any of `torch`'s higher-level capabilities -- not even *autograd*, its automatic-differentiation feature.
+
+This changed in the [follow-up post](https://blogs.rstudio.com/ai/posts/2020-10-05-torch-network-with-autograd). No more thinking about derivatives and the chain rule; a single call to `backward()` did it all.
+
+[Two days ago](https://blogs.rstudio.com/ai/posts/2020-10-07-torch-modules), the code again saw a major simplification. Modules, and compositions thereof, turned out to be much more manageable than tensors, hand-assembled in some DAG[^1].
+
+[^1]: directed acyclic graph
+
+Based on that last state, there are just two more things to do. For one, we still compute the loss by hand. And secondly, even though we get the gradients all nicely computed from *autograd*, we still loop over the model's parameters, updating them all ourselves. You won't be surprised to hear that nothing of this is necessary.
```

Comment:
> Based on that last state,

Does this mean you're picking up in the code where you left off in the third post? If not, that's kind of what it sounds like (since you're referencing "state").

Reply:
yeah, that's the intention - couldn't think of how to say this better :-(